Add DielectricFunction properties

JosePizarro3 opened the pull request:
Another property added to the list. In this case, there is a renaming to `Permittivity` to be in line with the taxonomy. I have to check if we can define `DielectricStrength`, and how to extract it from `Permittivity`.

The design idea here is that we want to store a 3x3 tensor $\epsilon(q, \omega)$. If it does not depend on $q$, we can extract the absorption spectrum as the imaginary term like $I(\omega) = Im \epsilon(\omega)$. Like the spectrum has a `rank=[]`, and the permittivity has `rank=[3, 3]`, I just needed to create 3 sub-sections for each principal direction for the spectrum.

I also:
- Added a new utils function to extract the specific list of variables (function `get_variables()`).
- Added abstraction for spectrum, defining `AbsorptionSpectrum` as a more abstract category for XAS, EELS. I added `axis` as a MEnum
- I had to add `KMesh` (from #71), and I also defined `Frequency` for variables
- Added `Permittivity.type` for simpler categorization

JosePizarro3 committed:
Improved numerical_settings schema for KSpace

Added KLinePath and KMesh variables with refs to the NumericalSettings sections

JosePizarro3 commented:
Hi @ndaelman-hu, @ladinesa 

I am not 100% who is better for reviewing this: I don't want comments on the schema, but rather on the implementation of functionalities and check that the testing is correct.

Let me know.

ndaelman-hu commented on src/nomad_simulations/utils/utils.py:
```diff
@@ -129,3 +129,26 @@ def is_not_representative(model_system, logger: BoundLogger):
+        return []
+    result = []
+    for var in variables:
+        if isinstance(var, variable_cls):
```
How specific does the type checking have to be? Should child classes also return `True`?

JosePizarro3 replied:
Well, this is defined by the `variable_cls`. Do you want me to be very specific and only the specific class (and not the inherited ones) is checked?

ndaelman-hu replied:
> Well, this is defined by the variable_cls

No, `isinstance` determines this, and it will return `True` for a parent, i.e. you will get `PotentialEnergy` sections / objects too when requesting `Energy`.

> Do you want me to be very specific and only the specific class (and not the inherited ones) is checked?

Idk, I was just asking whether this is the behavior you want. You can always add an optional toggle if you're unsure.

JosePizarro3 replied:
Yeah, it might be useful to specify only the class, and not treat inheritance or just use a toggle like you said. I will implement this.

JosePizarro3 replied:
Ok, see my comment below. I added a todo to tackle this later when we have an example of a variable which inherits from another variable.

ndaelman-hu commented on src/nomad_simulations/utils/utils.py:
```diff
@@ -129,3 +129,26 @@ def is_not_representative(model_system, logger: BoundLogger):
+
+
+# cannot define typing with `Variables` due to circular import issue
+def get_variables(
+    variables: Optional[List[ArchiveSection]], variable_cls: ArchiveSection
+) -> List[ArchiveSection]:
+    """
+    Get the list of variables which are of type `variable_cls` and appear under `variables`.
+
+    Args:
+        variables (List[Variables]): The list of variables to check.
+        variable_cls (Variables): The class of the variables to get.
+
+    Returns:
+        (List[Variables]): The list of variables which are of type `variable_cls`.
+    """
```
This approach is fine for small archives.
For larger, you'd likely want a specialized container class that holds an inverse index.

JosePizarro3 replied:
I am not sure if I understand the point of small archives. properties tend to be only of in terms of a few variables.

ndaelman-hu replied:
So, if I understood the suggested schema well, you will have 1 big array under `Output`, correct?
Up till now, you've only used `get_variables` for minor checks of individual properties. Do you foresee the need to search this property array as a whole?

JosePizarro3 replied:
> So, if I understood the suggested schema well, you will have 1 big array under Output, correct?

No. There is a list of properties as sub-sections. Under each property, we have to parse the variables as sub-sections which contain `points` and the property `value` with the shape coincide with the length of these variables (+ the rank defined for the property).

> Do you foresee the need to search this property array as a whole?

I don't understand this.

ndaelman-hu commented on tests/test_utils.py:
```diff
@@ -59,3 +65,16 @@ def test_is_not_representative():
 # ! Missing test for RusselSandersState (but this class will probably be deprecated)
+
+
+def test_get_variables():
```
Add a test for parent classes, e.g. `PotentialEnergy` inherits from `Energy`.
Check what happens when you request `Energy` then.

ndaelman-hu replied:
You should also add empty lists (and `None`, if you decide to support `Optional`).

Btw, I would run those as separate tests from this case, using parametrize: each test should look for a different failure mode.

JosePizarro3 replied:
I think for now I will keep it like this. I have not any case where this will happen, but for sure we will see it later. 

I added  a todo.

ndaelman-hu commented on tests/test_utils.py:
```diff
@@ -59,3 +65,16 @@ def test_is_not_representative():
+    ]
+    energies = get_variables(variables, Energy)
+    assert len(energies) == 1
+    assert np.allclose(energies[0].points.magnitude, [-3, -2, -1, 0, 1, 2, 3])
```
I think you can just check if the objects really match, i.e.
` assert energies == [variables[0]]`

ndaelman-hu commented on src/nomad_simulations/variables.py:
```diff
@@ -150,3 +152,76 @@ def __init__(
+        super().normalize(archive, logger)
+
+
+class KMesh(Variables):
+    """
+    K-point mesh over which the physical property is calculated. This is used to define `ElectronicEigenvalues(PhysicalProperty)` and
+    other k-space properties. The `points` are obtained from a refernece to the `NumericalSettings` section, `KMesh(NumericalSettings)`.
+    """
```
Wouldn't this be more appropriate in the KPoints PR?

JosePizarro3 replied:
Yeah, but while I am waiting, I just speed up the development in this branch by copy-pasting what is in there. Skip it and simply review it when we talk in #71 👍🏻

ndaelman-hu replied:
... You can just set a different target branch, then this won't show up.

JosePizarro3 replied:
But I don't want to show everything. Disregard it and that's it. This branch will be eventually rebased anyways.

ndaelman-hu commented on src/nomad_simulations/utils/utils.py:
```diff
@@ -129,3 +129,26 @@ def is_not_representative(model_system, logger: BoundLogger):
+    Returns:
+        (List[Variables]): The list of variables which are of type `variable_cls`.
+    """
+    if variables is None:
+        return []
```
The optional input is a bit non-sensical... You can just simplify this to a list and retain the logic below.

ndaelman-hu commented on src/nomad_simulations/properties/permittivity.py:
```diff
@@ -0,0 +1,118 @@
+            )
+            return None
+        # Extract the `Frequency` variable to extract the absorption spectrum
+        frequencies = get_variables(self.variables, Frequency)
```
If this is all `get_variables` is used for, I'd just use a one-liner, or define a quick lambda function. It hardly has the complexity to merit being a utility.

Moreover, I think you're restricting the usage scope of this function: it isn't limited to just `variables`. You could use the logic in other places too.

JosePizarro3 replied:
There is also some usage in spectral_profile.py. But we will use it in other situations.

This utility just returns the list of variables which matches the cls type passed. I will double-check metainfo.py to see whether this exists or not.

ndaelman-hu replied:
> This utility just returns the list of variables which matches the cls type passed.

Yes, I got that. That's why it's kind of trivial.

JosePizarro3 replied:
It is just a reusable function which will be probably used when parsing and normalizing PhysicalProperty.

ndaelman-hu reviewed (commented):
I only checked the `get_variables` utility, as you asked.
Check whether it's really worth having it as a utility and if so, the edge cases I mentioned.

I think this shows 1 weakness of the naive flat schema, is that we need these scans to provide lookup. For small archives, there's no real issue, but I'm unsure about search at the DB level for 2 reasons:

1. the schema does not serialize the actual class. Will we be able to read in more specialized child classes?
2. I don't know how an ES query will perform on these lists. Actually, do you have any tag to look for?

These 2 questions are more general than this PR. Just something th at I noticed.

ndaelman-hu commented:
> Hi @ndaelman-hu, @ladinesa
> 
> I am not 100% who is better for reviewing this: I don't want comments on the schema, but rather on the implementation of functionalities and check that the testing is correct.
> 
> Let me know.

Did a small review. Btw, you can also ask @Bernadette-Mohr. I think she's more than capable providing feedback on the utility.

JosePizarro3 commented:
> I only checked the `get_variables` utility, as you asked. Check whether it's really worth having it as a utility and if so, the edge cases I mentioned.
> 
> I think this shows 1 weakness of the naive flat schema, is that we need these scans to provide lookup. For small archives, there's no real issue, but I'm unsure about search at the DB level for 2 reasons:
> 
> 1. the schema does not serialize the actual class. Will we be able to read in more specialized child classes?
> 2. I don't know how an ES query will perform on these lists. Actually, do you have any tag to look for?
> 
> These 2 questions are more general than this PR. Just something th at I noticed.

Not sure if I understand the point of the weakness. This data structure only stores the data, and in case of talking about querying, I doubt searching for specific `variables` would ever be a topic (but rather, that a property has a variable_cls defined).

>  Actually, do you have any tag to look for?

Again, I am never talking about ES in this schema.

JosePizarro3 commented:
> > Hi @ndaelman-hu, @ladinesa
> > I am not 100% who is better for reviewing this: I don't want comments on the schema, but rather on the implementation of functionalities and check that the testing is correct.
> > Let me know.
> 
> Did a small review. Btw, you can also ask @Bernadette-Mohr. I think she's more than capable providing feedback on the utility.

Thanks! I will incorporate your feedback and merge. I think it makes sense if I actually merge after #71

ndaelman-hu commented:
> > I only checked the `get_variables` utility, as you asked. Check whether it's really worth having it as a utility and if so, the edge cases I mentioned.
> > I think this shows 1 weakness of the naive flat schema, is that we need these scans to provide lookup. For small archives, there's no real issue, but I'm unsure about search at the DB level for 2 reasons:
> > 
> > 1. the schema does not serialize the actual class. Will we be able to read in more specialized child classes?
> > 2. I don't know how an ES query will perform on these lists. Actually, do you have any tag to look for?
> > 
> > These 2 questions are more general than this PR. Just something th at I noticed.
> 
> Not sure if I understand the point of the weakness. This data structure only stores the data, and in case of talking about querying, I doubt searching for specific `variables` would ever be a topic (but rather, that a property has a variable_cls defined).
> 
> > Actually, do you have any tag to look for?
> 
> Again, I am never talking about ES in this schema.

You mentioned before essentially copying a filtered version of this schema over into `results`, that's why I'm asking.
If so, we should consider the points I raised, though we can discuss it outside of this PR. If searching for `variable_cls` is supported, my point 2 remains.

JosePizarro3 commented:
> > > I only checked the `get_variables` utility, as you asked. Check whether it's really worth having it as a utility and if so, the edge cases I mentioned.
> > > I think this shows 1 weakness of the naive flat schema, is that we need these scans to provide lookup. For small archives, there's no real issue, but I'm unsure about search at the DB level for 2 reasons:
> > > 
> > > 1. the schema does not serialize the actual class. Will we be able to read in more specialized child classes?
> > > 2. I don't know how an ES query will perform on these lists. Actually, do you have any tag to look for?
> > > 
> > > These 2 questions are more general than this PR. Just something th at I noticed.
> > 
> > 
> > Not sure if I understand the point of the weakness. This data structure only stores the data, and in case of talking about querying, I doubt searching for specific `variables` would ever be a topic (but rather, that a property has a variable_cls defined).
> > > Actually, do you have any tag to look for?
> > 
> > 
> > Again, I am never talking about ES in this schema.
> 
> You mentioned before essentially copying a filtered version of this schema over into `results`, that's why I'm asking. If so, we should consider the points I raised, though we can discuss it outside of this PR. If searching for `variable_cls` is supported, my point 2 remains.

I am still not sure. In `results` there will be a reference to the specific property (e.g., `permittivity`). The variables will never be searched, but probably it will be filtered to the result which variables we want to make searchable. This is as optimal as whatever is already in place for the `results` section when it is referencing properties in `run`.

Maybe I used wrong the word "copy"?

JosePizarro3 committed:
Squashing stuff to avoid issues when rebasing...

JosePizarro3 committed:
As anticipated, git is crap: now it created a function which was not existing in the previous commit when rebasing. what-the-actual-fuck ... -.-

JosePizarro3 commented:
I think I will be merging this tomorrow morning! No problem, we can revisit this as well in the future.

JosePizarro3 committed:
Improved get_variables utils function

JosePizarro3 committed:
Reduced test_permittivity.py for simplicity for debugging

JosePizarro3 committed:
Fix shape of Variables.point as it was producing some error when using refs in KMesh.points and KLinePath.points

Fixed test_permittivity

Added optional variables in get_variables utils function

JosePizarro3 committed:
Add small label in k_mesh_settings to clarify that is a q-mesh and this is its typical usage
